src/mappings/_utils.ts

```diff
@@ -127,6 +130,156 @@ export async function friendlyAmount(currency: Currency, amount: number): Promis
+                case 'KINT':
+                case 'KSM':
+                    return 12
+                case 'INTR':
+                case 'DOT':
+                    return 10
+                case 'KBTC':
+                case 'IBTC':
+                    return 8
+                default:
+                    return 0
+            }
+        case 'ForeignAsset':
+            const details = await getForeignAsset(currency.asset)
+            return details.decimals;
+        default:
+            return 0;
+    }
+}
+
+type CurrencyType = Bitcoin | Kintsugi | Kusama | Interlay | Polkadot;
+
+function mapCurrencyType(currency: Currency): CurrencyType {
+    switch(currency.isTypeOf) {
+        case 'NativeToken':
+            switch(currency.token) {
+                case 'KINT':
+                    return Kintsugi;
+                case 'KSM':
+                    return Kusama;
+                case 'INTR':
+                    return Interlay;
+                case 'DOT':
+                    return Polkadot;
+                case 'KBTC':
+                    return KBtc;
+                case 'INTR':
+                    return InterBtc;
+            }
+        case 'ForeignAsset':
+            return Bitcoin;
+        default:
+            throw new Error(`Unsupported currency type: ${currency.isTypeOf}`);
+    }
+}
+
+type OracleRate = {
+    btc: Big;
+    usdt: Big;
+}
+
+/* This function is used to calculate the exchange rate for a given currency at
+a given time.
+*/
+export async function getExchangeRate(
+    ctx: Ctx,
+    timestamp: number,
+    currency: Currency,
+    amount: number
+): Promise<OracleRate>  {
+    const mappedCurrency = mapCurrencyType(currency);
+    let baseMonetaryAmount
+    let searchBlock = currency.toJSON();
+
+    if (mappedCurrency === KBtc || mappedCurrency === InterBtc) {
+        baseMonetaryAmount = newMonetaryAmount(Big(1e8), Bitcoin)
+    }
+    else {
+        let lastUpdate = await ctx.store.get(OracleUpdate, {
+            where: { 
+                id: Like(`%${JSON.stringify(searchBlock)}`),
+                timestamp: LessThanOrEqual(new Date(timestamp)),
+            },
+            order: { timestamp: "DESC" },
+        });
+        console.log(lastUpdate?.timestamp);
+        if (lastUpdate === undefined) {
+            ctx.log.warn(
+                `WARNING: no price registered by Oracle for ${JSON.stringify(searchBlock)} at timestamp ${new Date(timestamp)}. Fetching first value.`
+            );
+            lastUpdate = await ctx.store.get(OracleUpdate, {
+                where: { 
+                    id: Like(`%${JSON.stringify(searchBlock)}`),
+                },
+                order: { timestamp: "ASC" },
+            });
+        }
+        const lastPrice = new Big((Number(lastUpdate?.updateValue) || 0) / 1e10);
+        baseMonetaryAmount = newMonetaryAmount(lastPrice, mappedCurrency);
+    }
+
+    searchBlock = {
+        isTypeOf: 'ForeignAsset',
+        asset: 1
```

Comment:
Are we sure the asset id will be the same for Kintsugi as well as Interlay?

Reply:
Philosophically -- no! Pragmatically, yes --- the asset has already been onboarded on both chains and changing the asset ID would be very hard (right?)

Reply:
I'd be more comfortable if we had a storage accessor for foreign assets (similar to [DexStablePoolsStorage](https://github.com/interlay/interbtc-squid/blob/d157f9cd76e7fab22c4a38a97568b4a752f20a9a/src/types/storage.ts#L5-L48)) where we can get all foreign assets and find (and cache) the one for USDT, eg. by filtering for `symbol === "USDT"`.

While I agree that Kintsugi and Interlay are extremely unlikely to ever change, I'm far from certain when it comes to the various testnets. If one of them is not initialized with foreign asset 1 being USDT it will lead to bugs that are very hard to figure out unless the person investigating knows about this hard-coded magic number.

Reply:
Raised as future issue: https://github.com/interlay/interbtc-squid/issues/97